provides.py

```diff
@@ -1,44 +1,30 @@
 from charmhelpers.core import hookenv
-from charms.reactive import hook
-from charms.reactive import RelationBase
-from charms.reactive import scopes
+from charms.reactive import when, when_not
+from charms.reactive import set_flag, clear_flag
+from charms.reactive import Endpoint
 
 
-class HttpProvides(RelationBase):
-    scope = scopes.GLOBAL
+class HttpProvides(Endpoint):
 
-    @hook('{provides:http}-relation-{joined,changed}')
-    def changed(self):
-        self.set_state('{relation_name}.available')
+    @when('endpoint.{endpoint_name}.joined')
+    def joined(self):
+        set_flag(self.expand_name('{endpoint_name}.available'))
 
-    @hook('{provides:http}-relation-departed')
-    def departed(self):
-        if len(self.conversation().units) == 1:
-            # this is the last departing unit
-            self.remove_state('{relation_name}.available')
+    @when_not('endpoint.{endpoint_name}.joined')
+    def broken(self):
+        clear_flag(self.expand_name('{endpoint_name}.available'))
 
-    def get_ingress_address(self):
-        try:
-            network_info = hookenv.network_get(self.relation_name)
-        except NotImplementedError:
-            network_info = []
-
-        if network_info and 'ingress-addresses' in network_info:
-            # just grab the first one for now, maybe be more robust here?
-            return network_info['ingress-addresses'][0]
-        else:
-            # if they don't have ingress-addresses they are running a juju that
-            # doesn't support spaces, so just return the private address
-            return hookenv.unit_get('private-address')
+    def get_ingress_address(self, rel_id=None):
+        # If no rel_id is provided, we fallback to the first one
+        if rel_id is None:
+            rel_id = self.relations[0].relation_id
+        return hookenv.ingress_address(rel_id, hookenv.local_unit())
 
     def configure(self, port, private_address=None, hostname=None):
-        if not hostname:
-            hostname = self.get_ingress_address()
-        if not private_address:
-            private_address = self.get_ingress_address()
-        relation_info = {
-            'hostname': hostname,
-            'private-address': private_address,
-            'port': port,
-        }
-        self.set_remote(**relation_info)
```

Comment:
`set_remote` is not an Endpoint method, yet some charms that implement this interface rely on it, e.g. graylog:

https://git.launchpad.net/graylog-charm/tree/reactive/graylog.py#n508

Is there a back compat way we can handle this either in the Endpoint class or on a per-interface level?

Reply:
If we don't handle `set_remote` at the Endpoints class level, I think we'll at least need to allow for charms that want to interact with reverse proxies like haproxy (which are not reactive and rely on `relation-get`), and for that, we'll need to provide more than the current host/ip/port.  Specifically, `services` relation data:

https://bazaar.launchpad.net/~haproxy-team/charm-haproxy/trunk/view/head:/hooks/hooks.py#L599

It doesn't have to be in `configure`. Maybe it could be in -- just spit ballin here -- a new `set_remote` func? ;)